Update laboratory1.sql

NastyaVicodin opened the pull request.

NastyaVicodin committed:
Update laboratory1.sql

ASIMER commented on km51/Lukianenko_Anastasiia/laboratory1.sql:
```diff
@@ -1,2 +1,127 @@
+
+ALTER TABLE PLAYLIST
+    ADD CONSTRAINT PLAYLIST_FK FOREIGN KEY MUSIC_NAME_FK REFERENCES MUSIC(MUSIC_NAME);
+
```
Отсутствует внешний ключ к таблице "HUMAN". Связь только с "MUSIC".
Название столбца основной таблицы нужно брать в скобки.
`ALTER TABLE PLAYLIST
    ADD CONSTRAINT PLAYLIST_FK FOREIGN KEY (MUSIC_NAME_FK) REFERENCES MUSIC(MUSIC_NAME);
ALTER TABLE PLAYLIST
    ADD CONSTRAINT PLAYLIST_FK FOREIGN KEY (HUMAN_NAME_FK) REFERENCES HUMAN(HUMAN_NAME);`

ASIMER commented on km51/Lukianenko_Anastasiia/laboratory1.sql:
```diff
@@ -1,2 +1,127 @@
+---------------------------------------------------------------------------*/
+--Код відповідь:
+
+GRANT CREATE ANY TABLE TO alukyanenko;
```
GRANT ALTER ANY TABLE TO alukyanenko;
GRANT DROP ANY TABLE TO alukyanenko;

ASIMER commented on km51/Lukianenko_Anastasiia/laboratory1.sql:
```diff
@@ -1,2 +1,127 @@
+
+ALTER TABLE PLAYLIST
+    ADD CONSTRAINT PLAYLIST_FK FOREIGN KEY MUSIC_NAME_FK REFERENCES MUSIC(MUSIC_NAME);
+
```
Отсутствует внешний ключ к таблице "HUMAN". Связь только с "MUSIC".
Название столбца основной таблицы нужно брать в скобки.
`ALTER TABLE PLAYLIST`
`    ADD CONSTRAINT PLAYLIST_FK1 FOREIGN KEY (MUSIC_NAME_FK) REFERENCES MUSIC(MUSIC_NAME);`
`ALTER TABLE PLAYLIST`
`    ADD CONSTRAINT PLAYLIST_FK2 FOREIGN KEY (HUMAN_NAME_FK) REFERENCES HUMAN(HUMAN_NAME);`

NastyaVicodin replied:
Відсутні додаткові поля та регулярні вирази до них. Кожна сутність має містити мінімум по 4 поля.
Неможна створювати зовнішній ключ із вже існуючим ім'ям.

ASIMER commented on km51/Lukianenko_Anastasiia/laboratory1.sql:
```diff
@@ -1,2 +1,127 @@
+--Код відповідь:
+
+GRANT CREATE ANY TABLE TO alukyanenko;
+
```
Еще не наданы права ALTER/GRANT
`GRANT ALTER ANY TABLE TO alukyanenko;`
`GRANT DROP ANY TABLE TO alukyanenko;`

NastyaVicodin replied:
ALTER та DROР дозволяють змінювати та видаляти таблицю відповідно.
Внесення даних - INSERT, яке було додано вище.
Виконання вибірки - SELECT.

NastyaVicodin replied:
ALTER та DROР дозволяють змінювати та видаляти таблицю відповідно.
Внесення даних - INSERT, яке було додано вище.
Виконання вибірки - SELECT.

ASIMER replied:
INSERT я увидел ранее потому не написал про него.
Да, нужно SELECT и не нужно то что я написал, я неправильно прочитал задание.

ASIMER commented on km51/Lukianenko_Anastasiia/laboratory1.sql:
```diff
@@ -1,2 +1,127 @@
+WHERE ORDERITEMS.ITEM_PRICE = MAX(ITEM_PRICE)
+    AND ORDERITEMS.PROD_ID = PRODUCTS.PROD_ID
+{PRODUCTS.PROD_NAME}
+
```
Я сделяль:
`PROJECT `
`ORDERITEMS TIMES PRODUCTS`
`
WHERE ORDERITEMS.ITEM_PRICE = MAX(PROJECT ORDERITEMS {ITEM_PRICE}) and
ORDERITEMS.PROD_ID = PRODUCTS.PROD_ID
`
`{PRODUCTS.prod_name}`

В SQL:
`SELECT DISTINCT PROD_NAME`
`from ORDERITEMS, PRODUCTS`
`WHERE ORDERITEMS.ITEM_PRICE in (SELECT MAX(ITEM_PRICE) from ORDERITEMS) and ORDERITEMS.PROD_ID = PRODUCTS.PROD_ID`

ASIMER commented on km51/Lukianenko_Anastasiia/laboratory1.sql:
```diff
@@ -1,2 +1,127 @@
+SELECT DISTINCT 
+    MIN(LENGTH(CUST_NAME)) AS "long_name"
+FROM CUSTOMERS;
+
```
В текущем коде оно выводит длину строки, но если добавить TRIM() оно выведет длину самой короткой строки.
Если переписать так, оно будет работать правильно:

`SELECT DISTINCT CUST_NAME AS "long_name"`
`FROM CUSTOMERS`
`WHERE LENGTH(TRIM(CUSTOMERS.CUST_NAME)) in (`
`    SELECT MIN(LENGTH(TRIM(CUSTOMERS.CUST_NAME)))`
`    FROM CUSTOMERS);`

ASIMER commented on km51/Lukianenko_Anastasiia/laboratory1.sql:
```diff
@@ -1,2 +1,127 @@
+    TRIM(CUST_NAME) || ' ' ||
+    TRIM(CUST_EMAIL) AS "client_name"
+FROM CUSTOMERS, ORDERS
+WHERE ORDERS.CUST_ID != CUSTOMERS.CUST_ID;
```
Данный код выводит пользователей когда их номер не найден в заказах.
Дисктинктом ты только убрала повторы но проблему не решила.
Но проблема в том что оно сравнивает номер пользователя с каждой строкой в таблице заказов, из за декартового произведения в FROM.
Нужно искать его в множестве заказов, что бы его там не было. Так не будет декартового произведения и оно просто переберет каждого пользователя по 1 разу, и так как они уникальны то повторов не будет и без фильтрации.

`SELECT TRIM(CUST_NAME)||' '|| TRIM(CUST_EMAIL) AS "client_name"`
`FROM CUSTOMERS`
`WHERE CUSTOMERS.CUST_ID not in (`
`    SELECT CUST_ID`
`    FROM ORDERS);`

NastyaVicodin committed:
Update laboratory2.sql

NastyaVicodin committed:
Update laboratory1.sql

1. Додані поля в старі таблиці та нова таблиця, встановлені зв'язки між таблицями.
2. Виправлені помилки та покращені запити в завданні 3.

NastyaVicodin reviewed (commented):
1. Додані поля в старі таблиці та нова таблиця, встановлені зв'язки між таблицями, додані CONSTRAINTS до кожного поля, кожна таблиця містить 3 рядка.
2. Виправлені помилки та покращені запити в завданні 3.
3. Додана оновлена версія таблиці, що згенерована у Power Designer.
4. Додані діаграми, побудовані в Power Designer.

NastyaVicodin committed:
Update laboratory1.sql

Вставлена інформація в таблиці

NastyaVicodin committed:
Update laboratory1.sql

CONSTRAINTS ADDED

NastyaVicodin committed:
Update laboratory1.sql

CONSTRAINTS edited.

NastyaVicodin commented:
![erd](https://user-images.githubusercontent.com/23039455/38773620-bf8aba4a-4059-11e8-8f12-bf2bf7ee7f20.jpg)

NastyaVicodin committed:
ERD

NastyaVicodin committed:
Added Power Designer version

NastyaVicodin committed:
Logical and Physical models added

NastyaVicodin committed:
Update laboratory3.sql

NastyaVicodin committed:
Update laboratory3.sql

NastyaVicodin committed:
Work with csv files

NastyaVicodin committed:
Oracle+Python+Plotly

Added new rows in the tables to make visualization better.

NastyaVicodin committed:
Update laboratory5.sql

NastyaVicodin committed:
Update laboratory4.sql
